### PR TITLE
F/WPC-153 comments listing route

### PR DESCRIPTION
**Pull request by bishoymelek**
**Comment List route** with:
- Comment details in Modal with all revisions timeline
- Sort (Modified at Column)
- Approve Action
- Reject Action with options of reason
- Request Changes Action with options of reason

### CODE REVIEW

File: packages/editor/src/client/routes/commentList.tsx [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+            </DescriptionListItem>
+            {confirmAction === ConfirmAction.Reject ||
+            confirmAction === ConfirmAction.RequestChanges ? (
+              <DescriptionListItem
+                label={t(
+                  confirmAction === ConfirmAction.Reject
+                    ? 'comments.panels.rejectionReason'
+                    : 'comments.panels.requestChangesReason'
+                )}>
+                <Dropdown
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:332]: Selected reasons are not shown
deepin-screen-recorder_Select area_20210208100647

File: packages/editor/src/client/routes/commentList.tsx [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+                </Dropdown>
+              </DescriptionListItem>
+            ) : null}
+          </DescriptionList>
+        </Modal.Body>
+        <Modal.Footer>
+          <Button
+            color={mapCommentActionToColor(confirmAction)}
+            disabled={isApproving || isRequestingChanges || isRejecting}
+            onClick={async () => {
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:360]: After confirming the action it should reload/refetch the data in the list. Refer to the `articleList` unpublish action for inspiration.

File: packages/editor/src/client/routes/commentList.tsx [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        overflow
+        onHide={() => setConfirmationDialogOpen(false)}>
+        <Modal.Header>
+          <Modal.Title>
+            <div>{t(mapModalTitle())}</div>
+          </Modal.Title>
+        </Modal.Header>
+        <Modal.Body>
+          <DescriptionList>
+            <DescriptionListItem label={t('comments.panels.userName')}>
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:298]: I want to see the id of the comment in the modal.

File: packages/editor/src/client/routes/commentList.tsx [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+                  : null}
+              </>
+            )}
+          </Cell>
+        </Column>
+        <Column width={150} align="left" resizable>
+          <HeaderCell>{t('comments.overview.userName')}</HeaderCell>
+          <Cell>{(rowData: CommentRefFragment) => <>{rowData.user?.name}</>}</Cell>
+        </Column>
+        <Column width={150} align="left" resizable>
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:198]: I want to see the title of the page or article in the table.

File: packages/editor/src/client/api/comment.graphql [hunk 5]
```diff
@@ -0,0 +1,71 @@
+fragment CommentRef on Comment {
```
Comment (roosnic1) [inline packages/editor/src/client/api/comment.graphql:1]: i would distinguish between `CommentRef` (id, state, revisions) and `FullComment` (everything)

File: packages/editor/src/client/routes/commentList.tsx [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+function mapCommentActionToColor(currentAction?: ConfirmAction) {
+  switch (currentAction) {
+    case ConfirmAction.Approve:
+      return 'green'
+    case ConfirmAction.RequestChanges:
+      return 'yellow'
+    case ConfirmAction.Reject:
+      return 'red'
+    default:
+      return 'green'
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:52]: bad practice. I'd return default color or none color if the type is unknown. Or make the parameter `currentAction` not optional i think then you don't need a default case.

File: packages/editor/src/client/routes/commentList.tsx [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+function mapCommentActionToTitle(currentAction?: ConfirmAction) {
+  switch (currentAction) {
+    case ConfirmAction.Approve:
+      return 'comments.panels.approve'
+    case ConfirmAction.RequestChanges:
+      return 'comments.panels.requestChanges'
+    case ConfirmAction.Reject:
+      return 'comments.panels.reject'
+    default:
+      return 'comments.panels.approve'
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:65]: Same here. Bad practice.

File: packages/editor/src/client/routes/commentList.tsx [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+  const [page, setPage] = useState(1)
+  const [limit, setLimit] = useState(10)
+  const [sortField, setSortField] = useState('createdAt')
+  const [sortOrder, setSortOrder] = useState<'asc' | 'desc'>('desc')
+
+  const [filter, setFilter] = useState('')
+
+  const [comments, setComments] = useState<CommentRefFragment[]>([])
+
+  const [approveComment, {loading: isApproving}] = useApproveCommentMutation()
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:91]: What about the error handling?

File: packages/editor/src/client/routes/commentList.tsx [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+  const [requestChanges, {loading: isRequestingChanges}] = useRequestChangesOnCommentMutation()
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:92]: What about the error handling?

File: packages/editor/src/client/routes/commentList.tsx [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+  const [rejectComment, {loading: isRejecting}] = useRejectCommentMutation()
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:93]: What about the error handling?

File: packages/editor/src/client/routes/commentList.tsx [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+      case ConfirmAction.Approve:
+        title = t('comments.panels.approveComment')
+        break
+      case ConfirmAction.Reject:
+        title = t('comments.panels.rejectComment')
+        break
+      case ConfirmAction.RequestChanges:
+        title = t('comments.panels.requestChangesOnComment')
+        break
+      default:
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:140]: You're working with enums. You shouldn't need defaults.

File: packages/editor/src/client/routes/commentList.tsx [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+          <InputGroup>
+            <Input value={filter} onChange={value => setFilter(value)} />
+            <InputGroup.Addon>
+              <Icon icon="search" />
+            </InputGroup.Addon>
+          </InputGroup>
+        </FlexboxGrid.Item>
+      </FlexboxGrid>
+
+      <Table
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:163]: Checkout https://github.com/wepublish/wepublish/pull/172/files#diff-6efbad89c5a8f1d9efb955a581e41cc41b43726e032b3aa53a2f67bfc3437a69R106 

with the extra div and the min height you can avoid a bug where the autoHeight sometimes cuts of the last entry.

File: packages/editor/src/client/routes/commentList.tsx [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+                case CommentState.PendingApproval:
+                  state = 'comments.state.pendingApproval'
+                  break
+                case CommentState.PendingUserChanges:
+                  state = 'comments.state.pendingUserChanges'
+                  break
+                case CommentState.Rejected:
+                  state = 'comments.state.rejected'
+                  break
+                default:
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:216]: no defaults with enums

File: packages/editor/src/client/routes/commentList.tsx [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+                case ConfirmAction.Approve:
+                  await approveComment({
+                    variables: {
+                      id: currentComment.id
+                    }
+                  })
+                  setConfirmationDialogOpen(false)
+                  break
+                case ConfirmAction.RequestChanges:
+                  if (!rejectionReason) {
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:372]: Disable the confirm button if there is no rejectionReason. It's bad UX if you allow an action to happen if the data is not complete.

**Review by roosnic1 — CHANGES_REQUESTED** [review]
Looks very good but i found some places for improvement.

File: packages/editor/src/client/routes/commentList.tsx (hunk 2 above)
Comment (bishoymelek) [reply packages/editor/src/client/routes/commentList.tsx:364]: I couldn't understand when the re-fetch happens in articleList route but I did it my way, please let me know your opinion

**Review by penk85 — APPROVED** [review]

File: packages/editor/src/client/api/comment.graphql (hunk 5 above)
Comment (bishoymelek) [reply packages/editor/src/client/api/comment.graphql]: @roosnic1  When to use CommentRef?

Comment (roosnic1) [reply packages/editor/src/client/api/comment.graphql]: maybe not now but for the future.

File: packages/editor/src/client/routes/commentList.tsx [hunk 15]
```diff
@@ -0,0 +1,501 @@
...
+                    resetCurrentCommentState()
+                    break
+                  case ConfirmAction.Reject:
+                    if (!rejectionReason) return
+                    await rejectComment({
+                      variables: {
+                        id: currentComment.id,
+                        rejectionReason
+                      },
+                      update: cache => {
```
Comment (roosnic1) [inline packages/editor/src/client/routes/commentList.tsx:459]: Try not to repeat yourself. The update function seems to always do exactly the same.  I think you can move that into a function.

**Review by roosnic1 — CHANGES_REQUESTED** [review]
looks good just a tiny improvement

**Review by roosnic1 — APPROVED** [review]
👍🏾

**Review by 0815pascal — APPROVED** [review]
Looks good 👍🏼

**Review by HanaAyman — APPROVED** [review]

**Pull request merged**